File: go/coglog.go
// Package coglog provides CogLog — minimal cognitive continuity for LLMs.
//
// A single-window (size 1) log that holds the previous turn's three-layer
// structure (user utterance, thinking process, assistant output) plus a
// four-axis interpretation layer (current_focus, theory_of_mind,
// self_narrative, annotation), making it available at the start of the
// next turn.
package coglog

import (
	"encoding/json"
	"errors"
	"fmt"
	"os"
	"path/filepath"
	"time"
)

// ═══════════════════════════════════════════════════════════════════
// Data types
// ═══════════════════════════════════════════════════════════════════

// FactLayerSchema describes the fact layer in _schema.
type FactLayerSchema struct {
	User      string `json:"user"`
	Thinking  string `json:"thinking"`
	Assistant string `json:"assistant"`
}

// InterpretationLayerSchema describes the interpretation layer in _schema.
type InterpretationLayerSchema struct {
	CurrentFocus  string `json:"current_focus"`
	TheoryOfMind  string `json:"theory_of_mind"`
	SelfNarrative string `json:"self_narrative"`
	Annotation    string `json:"annotation"`
}

// ConstraintsSchema describes constraints in _schema.
type ConstraintsSchema struct {
	WindowSize          string `json:"window_size"`
	InterpretationEmpty string `json:"interpretation_empty"`
}

// Schema is the self-documenting _schema field embedded in each entry.
type Schema struct {
	Version             string                    `json:"version"`
	FactLayer           FactLayerSchema           `json:"fact_layer"`
	InterpretationLayer InterpretationLayerSchema `json:"interpretation_layer"`
	Constraints         ConstraintsSchema         `json:"constraints"`
}

// Layers holds the fact layer: what happened.
type Layers struct {
	User      string `json:"user"`
	Thinking  string `json:"thinking"`
	Assistant string `json:"assistant"`
}

// Entry is a single CogLog turn.
type Entry struct {
	Schema       Schema `json:"_schema"`
	TurnID       int    `json:"turn_id"`
	Timestamp    string `json:"timestamp"`
	Layers       Layers `json:"layers"`
	CurrentFocus  string `json:"current_focus"`
	TheoryOfMind  string `json:"theory_of_mind"`
	SelfNarrative string `json:"self_narrative"`
	Annotation    string `json:"annotation"`
}

// WriteArgs is the input for writing a new entry.
type WriteArgs struct {
	User          string `json:"user"`
	Thinking      string `json:"thinking"`
	Assistant     string `json:"assistant"`
	CurrentFocus  string `json:"current_focus"`
	TheoryOfMind  string `json:"theory_of_mind"`
	SelfNarrative string `json:"self_narrative"`
	Annotation    string `json:"annotation"`
}

// ClearResult is the result of a clear operation.
type ClearResult struct {
	Cleared bool   `json:"cleared"`
	Reason  string `json:"reason,omitempty"`
}

// ═══════════════════════════════════════════════════════════════════
// Schema constant
// ═══════════════════════════════════════════════════════════════════

// MakeSchema returns the CogLog schema.
func MakeSchema() Schema {
	return Schema{
		Version: "0.9.1",  // @coglog-version
		FactLayer: FactLayerSchema{
			User:      "non-empty string required \u2014 user's original utterance",
			Thinking:  "non-empty string required \u2014 AI's full thinking process",
			Assistant: "non-empty string required \u2014 AI's original output",
		},
		InterpretationLayer: InterpretationLayerSchema{
			CurrentFocus:  "string required, empty OK \u2014 present: what am I working on?",
			TheoryOfMind:  "string required, empty OK \u2014 other: what is the user's state?",
			SelfNarrative: "string required, empty OK \u2014 self: who am I in this moment?",
			Annotation:    "string required, empty OK \u2014 future: what should I do next?",
		},
		Constraints: ConstraintsSchema{
			WindowSize:          "1 turn (overwritten each write)",
			InterpretationEmpty: "choosing not to write is itself a metacognitive act",
		},
	}
}

// ═══════════════════════════════════════════════════════════════════
// Validation
// ═══════════════════════════════════════════════════════════════════

// ErrValidation indicates a validation failure.
var ErrValidation = errors.New("validation error")

// Validate checks that WriteArgs satisfy CogLog constraints.
// Fact layer fields must be non-empty strings.
// Interpretation layer fields are strings (empty acceptable).
func (a *WriteArgs) Validate() error {
	if a.User == "" {
		return fmt.Errorf("%w: missing required field: user", ErrValidation)
	}
	if a.Thinking == "" {
		return fmt.Errorf("%w: missing required field: thinking", ErrValidation)
	}
	if a.Assistant == "" {
		return fmt.Errorf("%w: missing required field: assistant", ErrValidation)
	}
	return nil
}

// ParseWriteArgs parses JSON bytes into WriteArgs, verifying that all
// interpretation layer fields are explicitly present in the input.
// Go's json.Unmarshal silently assigns zero values to missing fields,
// so we first check field presence via map[string]interface{}.
func ParseWriteArgs(data []byte) (WriteArgs, error) {
	var raw map[string]interface{}
	if err := json.Unmarshal(data, &raw); err != nil {
		return WriteArgs{}, fmt.Errorf("invalid JSON: %w", err)
	}
	for _, key := range []string{"current_focus", "theory_of_mind", "self_narrative", "annotation"} {
		val, ok := raw[key]
		if !ok {
			return WriteArgs{}, fmt.Errorf("%w: missing required field: %s (empty string is acceptable)", ErrValidation, key)
		}
		if _, isStr := val.(string); !isStr {
			return WriteArgs{}, fmt.Errorf("%w: field %s must be a string", ErrValidation, key)
		}
	}
	var args WriteArgs
	if err := json.Unmarshal(data, &args); err != nil {
		return WriteArgs{}, fmt.Errorf("invalid JSON: %w", err)
	}
	return args, nil
}

// ═══════════════════════════════════════════════════════════════════
// Advance — pure entry construction
// ═══════════════════════════════════════════════════════════════════

// Advance constructs a new Entry from the previous entry and write arguments.
// This is the core recurrence relation: a_{n+1} = f(a_n, x_n).
func Advance(prev *Entry, args WriteArgs, now time.Time) Entry {
	turnID := 1
	if prev != nil {
		turnID = prev.TurnID + 1
	}
	return Entry{
		Schema:    MakeSchema(),
		TurnID:    turnID,
		Timestamp: now.UTC().Format("2006-01-02T15:04:05Z"),
		Layers: Layers{
			User:      args.User,
			Thinking:  args.Thinking,
			Assistant: args.Assistant,
		},
		CurrentFocus:  args.CurrentFocus,
		TheoryOfMind:  args.TheoryOfMind,
		SelfNarrative: args.SelfNarrative,
		Annotation:    args.Annotation,
	}
}

// ═══════════════════════════════════════════════════════════════════
// CogLog — file I/O layer
// ═══════════════════════════════════════════════════════════════════

// CogLog manages a single-window coglog stored as current.json.
type CogLog struct {
	DataDir     string
	CurrentFile string
}

// defaultCoglogDir returns the default data directory.
// Priority: COGLOG_DIR env > $HOME/.coglog > ./.coglog (final fallback)
func defaultCoglogDir() string {
	if d := os.Getenv("COGLOG_DIR"); d != "" {
		return d
	}
	home, err := os.UserHomeDir()
	if err != nil {
		return ".coglog"
	}
	return filepath.Join(home, ".coglog")
}

// New creates a CogLog with the default data directory (~/.coglog).
func New() *CogLog {
	dataDir := defaultCoglogDir()
	return &CogLog{
		DataDir:     dataDir,
		CurrentFile: filepath.Join(dataDir, "current.json"),
	}
}

// NewWithDir creates a CogLog with a custom data directory.
func NewWithDir(dir string) *CogLog {
	return &CogLog{
		DataDir:     dir,
		CurrentFile: filepath.Join(dir, "current.json"),
	}
}

// Read returns the previous turn's entry, or nil if none exists.
func (c *CogLog) Read() (*Entry, error) {
	data, err := os.ReadFile(c.CurrentFile)
	if err != nil {
		if os.IsNotExist(err) {
			return nil, nil
		}
		return nil, err
	}
	var entry Entry
	if err := json.Unmarshal(data, &entry); err != nil {
		return nil, err
	}
	return &entry, nil
}

// Write validates args, constructs a new entry, and writes it to current.json.
func (c *CogLog) Write(args WriteArgs) (*Entry, error) {
	if err := args.Validate(); err != nil {
		return nil, err
	}
	if err := os.MkdirAll(c.DataDir, 0755); err != nil {
		return nil, err
	}
	prev, err := c.Read()
	if err != nil {
		return nil, err
	}
	entry := Advance(prev, args, time.Now())
	data, err := json.MarshalIndent(entry, "", "  ")
	if err != nil {
		return nil, err
	}
	data = append(data, '\n')
	if err := os.WriteFile(c.CurrentFile, data, 0644); err != nil {
		return nil, err
	}
	return &entry, nil
}

// Clear removes the current.json file.
func (c *CogLog) Clear() ClearResult {
	if err := os.Remove(c.CurrentFile); err != nil {
		if os.IsNotExist(err) {
			return ClearResult{Cleared: false, Reason: "no existing coglog"}
		}
		return ClearResult{Cleared: false, Reason: err.Error()}
	}
	return ClearResult{Cleared: true}
}

File: ruby/coglog/lib/coglog.rb
# frozen_string_literal: true

# CogLog — Minimal cognitive continuity for LLMs.
#
# A single-window (size 1) log that holds the previous turn's three-layer
# structure (user utterance, thinking process, assistant output) plus a
# four-axis interpretation layer (current_focus, theory_of_mind,
# self_narrative, annotation), making it available at the start of the
# next turn.

require 'json'
require 'time'
require 'fileutils'

module CogLog
  VERSION = '0.9.1'  # @coglog-version

  SCHEMA = {
    'version' => VERSION,
    'fact_layer' => {
      'user'      => "non-empty string required \u2014 user's original utterance",
      'thinking'  => "non-empty string required \u2014 AI's full thinking process",
      'assistant' => "non-empty string required \u2014 AI's original output"
    },
    'interpretation_layer' => {
      'current_focus'  => "string required, empty OK \u2014 present: what am I working on?",
      'theory_of_mind' => "string required, empty OK \u2014 other: what is the user's state?",
      'self_narrative'  => "string required, empty OK \u2014 self: who am I in this moment?",
      'annotation'     => "string required, empty OK \u2014 future: what should I do next?"
    },
    'constraints' => {
      'window_size'          => '1 turn (overwritten each write)',
      'interpretation_empty' => 'choosing not to write is itself a metacognitive act'
    }
  }.freeze

  FACT_KEYS = %w[user thinking assistant].freeze
  INTERP_KEYS = %w[current_focus theory_of_mind self_narrative annotation].freeze

  class ValidationError < StandardError; end

  class CogLog
    attr_reader :data_dir, :current_file

    def initialize(data_dir: nil)
      env = ENV['COGLOG_DIR']
      env = nil if env&.empty?
      @data_dir = data_dir || env || File.join(Dir.home, '.coglog')
      @current_file = File.join(@data_dir, 'current.json')
    end

    def read
      return nil unless File.exist?(@current_file)

      JSON.parse(File.read(@current_file, encoding: 'UTF-8'))
    end

    def write(user:, thinking:, assistant:, current_focus:, theory_of_mind:, self_narrative:, annotation:)
      # Fact layer: required, non-empty
      { 'user' => user, 'thinking' => thinking, 'assistant' => assistant }.each do |key, val|
        unless val.is_a?(String) && !val.empty?
          raise ValidationError, "missing required field: #{key}"
        end
      end

      # Interpretation layer: required, empty string acceptable
      { 'current_focus' => current_focus, 'theory_of_mind' => theory_of_mind,
        'self_narrative' => self_narrative, 'annotation' => annotation }.each do |key, val|
        unless val.is_a?(String)
          raise ValidationError, "missing required field: #{key} (empty string is acceptable)"
        end
      end

      FileUtils.mkdir_p(@data_dir)

      prev = self.read
      turn_id = prev ? prev['turn_id'] + 1 : 1

      entry = {
        '_schema'        => SCHEMA,
        'turn_id'        => turn_id,
        'timestamp'      => Time.now.utc.strftime('%Y-%m-%dT%H:%M:%SZ'),
        'layers'         => { 'user' => user, 'thinking' => thinking, 'assistant' => assistant },
        'current_focus'  => current_focus,
        'theory_of_mind' => theory_of_mind,
        'self_narrative'  => self_narrative,
        'annotation'     => annotation
      }

      File.write(@current_file, JSON.pretty_generate(entry) + "\n", encoding: 'UTF-8')
      entry
    end

    def clear
      if File.exist?(@current_file)
        File.delete(@current_file)
        { 'cleared' => true }
      else
        { 'cleared' => false, 'reason' => 'no existing coglog' }
      end
    end
  end

  # ── CLI ──────────────────────────────────────────────────────────

  USAGE = <<~TEXT
    usage: coglog-cli <read|write|clear>

      read    — display the previous turn's coglog
      write   — save current turn (reads JSON from stdin)
      clear   — reset coglog

    write expects JSON on stdin (all fields required):
      {
        "user": "user's message",
        "thinking": "AI thinking process",
        "assistant": "AI output",
        "current_focus": "what is happening right now",
        "theory_of_mind": "user intent/state inference",
        "self_narrative": "improvised self-story at this moment",
        "annotation": "note to future self"
      }

      fact layer (user, thinking, assistant): non-empty string required
      interpretation layer (current_focus, theory_of_mind,
        self_narrative, annotation): string required, empty string acceptable
  TEXT

  def self.cli(args = ARGV)
    # Parse --coglog-dir <path> (priority: arg > COGLOG_DIR env > default)
    args = args.dup
    coglog_dir = nil
    if args[0] == '--coglog-dir' && args[1]
      coglog_dir = args[1]
      args.shift(2)
    end
    ml = CogLog.new(data_dir: coglog_dir)
    command = args[0]

    case command
    when 'read'
      entry = ml.read
      if entry.nil?
        puts '(no coglog found)'
      else
        puts JSON.pretty_generate(entry)
      end

    when 'write'
      data = JSON.parse($stdin.read)
      entry = ml.write(
        user:          data['user'],
        thinking:      data['thinking'],
        assistant:     data['assistant'],
        current_focus: data['current_focus'],
        theory_of_mind: data['theory_of_mind'],
        self_narrative: data['self_narrative'],
        annotation:    data['annotation']
      )
      puts "coglog: turn #{entry['turn_id']} written"

    when 'clear'
      result = ml.clear
      if result['cleared']
        puts 'coglog: cleared'
      else
        puts "coglog: #{result['reason']}"
      end

    else
      print USAGE
      exit(command ? 1 : 0)
    end
  rescue ValidationError => e
    warn "coglog error: #{e.message}"
    exit 1
  rescue JSON::ParserError => e
    warn "coglog error: invalid JSON: #{e.message}"
    exit 1
  rescue StandardError => e
    warn "coglog error: #{e.message}"
    exit 1
  end
end
